File: src/scrapers/decathlon_ch.py
"""Scraper for Decathlon.ch Blackroll brand page."""

from __future__ import annotations

import asyncio
import re
from datetime import UTC, datetime
from urllib.parse import urljoin

from playwright.async_api import Page

from ..models import Product, ScrapeResult
from ..utils import parse_price
from .base import BaseScraper
from .browser_pool import get_browser_context

_DESKTOP_UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
)

_STEALTH_INIT_SCRIPT = """
Object.defineProperty(navigator, 'webdriver', { get: () => undefined });
Object.defineProperty(navigator, 'languages', { get: () => ['de-CH', 'de', 'en'] });
Object.defineProperty(navigator, 'plugins', { get: () => [1, 2, 3, 4, 5] });
""".strip()


class DecathlonCHScraper(BaseScraper):
    name = "decathlon_ch"
    display_name = "Decathlon.ch (Blackroll)"
    url = "https://www.decathlon.ch/de/brands/blackroll"

    user_agent = _DESKTOP_UA
    locale = "de-CH"
    viewport = {"width": 1920, "height": 1080}
    init_scripts = [_STEALTH_INIT_SCRIPT]

    _base_url = "https://www.decathlon.ch"
    _default_currency = "CHF"

    async def scrape(self) -> ScrapeResult:
        async with get_browser_context(
            browser_type=self.browser_type,
            user_agent=self.user_agent,
            locale=self.locale,
            viewport=self.viewport,
            init_scripts=self.init_scripts,
        ) as context:
            page = await context.new_page()
            print(f"[{self.name}] Loading {self.url}...")
            await page.goto(self.url, wait_until="domcontentloaded", timeout=90000)
            await self._handle_cookie_consent(page)
            await self._wait_for_products(page)
            products = await self.extract_products(page)

        deduped: dict[str, Product] = {}
        for product in products:
            key = product.item_id or product.url or product.name
            if not key:
                continue
            deduped[key] = product

        final_products = list(deduped.values())
        print(f"[{self.name}] Extracted {len(final_products)} products")

        return ScrapeResult(
            source=self.name,
            source_url=self.url,
            scraped_at=datetime.now(UTC),
            products=final_products,
        )

    async def _handle_cookie_consent(self, page: Page) -> None:
        for selector in [
            "#didomi-notice-agree-button",
            'button:has-text("Annehmen und Schliessen")',
            'button:has-text("Accept all")',
            'button:has-text("Accept")',
            'button:has-text("Alle akzeptieren")',
            'button:has-text("Allem zustimmen")',
            "#onetrust-accept-btn-handler",
            '[data-testid="uc-accept-all-button"]',
        ]:
            try:
                btn = await page.query_selector(selector)
                if btn and await btn.is_visible():
                    await btn.click()
                    await page.wait_for_timeout(600)
                    return
            except Exception:
                continue

    async def _wait_for_products(self, page: Page) -> None:
        await page.wait_for_selector('article:has(a[href*="/p/"])', timeout=60000, state="attached")
        await page.wait_for_timeout(1200)

    @staticmethod
    def _pick_srcset_url(srcset: str | None) -> str | None:
        if not srcset or not isinstance(srcset, str):
            return None
        candidates: list[tuple[int, str]] = []
        for part in [p.strip() for p in srcset.split(",") if p.strip()]:
            chunks = part.split()
            if not chunks:
                continue
            url = chunks[0].strip()
            if not url:
                continue
            score = 0
            if len(chunks) >= 2:
                hint = chunks[1].strip().lower()
                if hint.endswith("w"):
                    try:
                        score = int(re.sub(r"[^0-9]", "", hint) or "0")
                    except Exception:
                        score = 0
                elif hint.endswith("x"):
                    try:
                        score = int(float(hint[:-1]) * 1000)
                    except Exception:
                        score = 0
            candidates.append((score, url))
        if not candidates:
            return None
        candidates.sort(key=lambda t: t[0])
        return candidates[-1][1]

    @staticmethod
    def _extract_item_id(product_url: str | None) -> str | None:
        if not product_url:
            return None
        if match := re.search(r"/R-p-([0-9a-fA-F-]{8,})", product_url):
            return match.group(1)
        if match := re.search(r"[?&]mc=([^&#]+)", product_url):
            return match.group(1)
        return None

    def _normalize_url(self, value: str | None) -> str | None:
        if not value or not isinstance(value, str):
            return None
        url = value.strip()
        if not url:
            return None
        if url.startswith("//"):
            return "https:" + url
        if url.startswith("/"):
            return urljoin(self._base_url, url)
        return url

    async def _fetch_image(
        self, page: Page, image_url: str | None, *, referer: str
    ) -> tuple[bytes | None, str | None]:
        image_url = self._normalize_url(image_url)
        if not image_url:
            return None, None

        try:
            resp = await page.context.request.get(
                image_url,
                headers={
                    # Avoid requesting AVIF; Pillow cannot reliably decode it here.
                    "Accept": "image/webp,image/apng,image/*,*/*;q=0.8",
                    "Referer": referer,
                    "Sec-Fetch-Dest": "image",
                    "Sec-Fetch-Mode": "no-cors",
                },
                timeout=30000,
            )
            if not resp.ok:
                return None, None
            body = await resp.body()
            if len(body) < 800:
                return None, None
            mime = resp.headers.get("content-type")
            if mime and ";" in mime:
                mime = mime.split(";", 1)[0].strip()
            return body, mime
        except Exception:
            return None, None

    async def extract_products(self, page: Page) -> list[Product]:
        items = await page.eval_on_selector_all(
            'article:has(a[href*="/p/"])',
            """
            els => els.map(el => {
              const a = el.querySelector('a[href*="/p/"]');
              const img = el.querySelector('img');
              const priceEl = el.querySelector('.vp-price [data-part="amount"], [data-part="amount"]');
              const url = a ? a.href : null;
              const nameFromImg = img && img.alt ? img.alt.trim() : null;
              const nameFromLink = a && a.textContent ? a.textContent.trim() : null;
              const srcset = img ? (img.getAttribute('srcset') || img.getAttribute('data-srcset')) : null;
              const imgUrl = img ? (img.currentSrc || img.src || img.getAttribute('data-src')) : null;
              const priceText = priceEl ? priceEl.textContent.trim() : null;
              return { url, name: nameFromImg || nameFromLink, priceText, imgUrl, srcset };\n            })
            """,
        )
        if not isinstance(items, list) or not items:
            return []

        parsed: list[dict] = []
        for raw in items:
            if not isinstance(raw, dict):
                continue
            product_url = raw.get("url")
            name = (raw.get("name") or "").strip() if isinstance(raw.get("name"), str) else ""
            if not product_url or not isinstance(product_url, str) or not name:
                continue
            clean_url = product_url.split("?", 1)[0].strip()
            if not clean_url:
                continue

            image_url = None
            srcset = raw.get("srcset") if isinstance(raw.get("srcset"), str) else None
            if srcset:
                image_url = self._pick_srcset_url(srcset)
            if not image_url:
                image_url = raw.get("imgUrl") if isinstance(raw.get("imgUrl"), str) else None

            price_text = raw.get("priceText") if isinstance(raw.get("priceText"), str) else None
            amount, currency = parse_price(price_text)
            parsed.append(
                {
                    "name": name,
                    "url": clean_url,
                    "item_id": self._extract_item_id(product_url) or clean_url,
                    "price": amount,
                    "currency": currency or self._default_currency,
                    "image_url": image_url,
                }
            )

        semaphore = asyncio.Semaphore(6)

        async def build_product(entry: dict) -> Product | None:
            async with semaphore:
                image_bytes, image_mime = await self._fetch_image(
                    page, entry.get("image_url"), referer=entry["url"]
                )
            return Product(
                name=entry["name"],
                price=entry.get("price"),
                currency=entry.get("currency"),
                url=entry.get("url"),
                item_id=entry.get("item_id"),
                image=image_bytes,
                image_mime=image_mime,
            )

        products = await asyncio.gather(*(build_product(e) for e in parsed))
        return [p for p in products if p is not None]

File: new/packages/scrapers/src/decathlonCh.ts
import type { DiscoveredProduct } from "@corastuff/shared";
import type { Page } from "playwright";
import { withPlaywrightContext, type PlaywrightContextProfile } from "./playwrightContext.js";

function asNonEmptyString(value: unknown): string | undefined {
  return typeof value === "string" && value.trim() ? value.trim() : undefined;
}

function normalizeUrl(baseUrl: string, value: string | undefined): string | undefined {
  const raw = asNonEmptyString(value);
  if (!raw) return undefined;
  if (raw.startsWith("//")) return `https:${raw}`;
  if (raw.startsWith("/")) return new URL(raw, baseUrl).toString();
  try {
    return new URL(raw, baseUrl).toString();
  } catch {
    return raw;
  }
}

function normalizeSpaces(value: string) {
  return value.replace(/\s+/g, " ").trim();
}

function pickBestSrcsetUrl(srcset: string | null | undefined): string | undefined {
  const raw = asNonEmptyString(srcset ?? undefined);
  if (!raw) return undefined;

  const candidates: Array<{ score: number; url: string }> = [];
  for (const part of raw.split(",")) {
    const chunk = part.trim();
    if (!chunk) continue;
    const pieces = chunk.split(/\s+/);
    const url = asNonEmptyString(pieces[0]);
    if (!url) continue;

    const hint = asNonEmptyString(pieces[1]);
    let score = 0;
    if (hint) {
      const lower = hint.toLowerCase();
      if (lower.endsWith("w")) {
        const num = Number.parseInt(lower.replace(/[^\d]/g, ""), 10);
        score = Number.isFinite(num) ? num : 0;
      } else if (lower.endsWith("x")) {
        const mult = Number.parseFloat(lower.replace(/x$/, ""));
        score = Number.isFinite(mult) ? Math.round(mult * 1000) : 0;
      }
    }

    candidates.push({ score, url });
  }

  if (candidates.length === 0) return undefined;
  candidates.sort((a, b) => a.score - b.score);
  return candidates[candidates.length - 1]!.url;
}

function extractItemId(productUrl: string): string | undefined {
  try {
    const u = new URL(productUrl);
    const mc = asNonEmptyString(u.searchParams.get("mc") ?? undefined);
    if (mc) return mc;
    const match = u.pathname.match(/\/R-p-([0-9a-fA-F-]{8,})/);
    if (match?.[1]) return match[1];
  } catch {
    // ignore
  }
  return undefined;
}

function parsePriceText(text: string | undefined): { price?: number; currency?: string } {
  const raw = asNonEmptyString(text);
  if (!raw) return {};

  const currency = raw.includes("€") ? "EUR" : raw.toUpperCase().includes("CHF") ? "CHF" : undefined;
  const cleaned = raw.replace(/[^\d,.\-]/g, " ").replace(/\s+/g, " ").trim();
  const match = cleaned.match(/-?\d[\d.,]*/);
  if (!match?.[0]) return { ...(currency ? { currency } : {}) };

  const token = match[0];
  const lastComma = token.lastIndexOf(",");
  const lastDot = token.lastIndexOf(".");

  const normalized = (() => {
    if (lastComma !== -1 && lastDot !== -1) {
      if (lastComma > lastDot) return token.replace(/\./g, "").replace(",", ".");
      return token.replace(/,/g, "");
    }
    if (lastComma !== -1) {
      const fracLen = token.length - lastComma - 1;
      if (fracLen === 2) return token.replace(/\./g, "").replace(",", ".");
      return token.replace(/,/g, "");
    }
    if (lastDot !== -1) {
      const fracLen = token.length - lastDot - 1;
      if (fracLen === 2) return token.replace(/,/g, "");
      return token.replace(/\./g, "");
    }
    return token;
  })();

  const num = Number.parseFloat(normalized);
  if (!Number.isFinite(num)) return { ...(currency ? { currency } : {}) };
  return { price: num, ...(currency ? { currency } : {}) };
}

async function clickFirstVisible(page: Page, selectors: string[]) {
  for (const selector of selectors) {
    try {
      const loc = page.locator(selector).first();
      if (await loc.isVisible({ timeout: 1500 })) {
        await loc.click({ timeout: 5000 });
        await page.waitForTimeout(600);
        return true;
      }
    } catch {
      // ignore
    }
  }
  return false;
}

async function handleCookieConsent(page: Page) {
  const selectors = [
    "#didomi-notice-agree-button",
    'button:has-text("Annehmen und Schliessen")',
    'button:has-text("Accept all")',
    'button:has-text("Accept")',
    'button:has-text("Alle akzeptieren")',
    'button:has-text("Allem zustimmen")',
    "#onetrust-accept-btn-handler",
    '[data-testid="uc-accept-all-button"]'
  ];

  if (await clickFirstVisible(page, selectors)) return;
  for (const frame of page.frames()) {
    for (const selector of selectors) {
      try {
        const loc = frame.locator(selector).first();
        if (await loc.isVisible({ timeout: 1500 })) {
          await loc.click({ timeout: 5000 });
          await page.waitForTimeout(600);
          return;
        }
      } catch {
        // ignore
      }
    }
  }
}

const desktopUserAgent =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36";

export async function scrapeDecathlonChBrandPage(options: {
  sourceSlug: string;
  listingUrl: string;
  baseUrl?: string;
  currency?: string;
  browser?: PlaywrightContextProfile;
  userAgent?: string;
  locale?: string;
  log?: (message: string) => void;
}): Promise<{ products: DiscoveredProduct[] }> {
  const baseUrl = options.baseUrl ?? new URL(options.listingUrl).origin;
  const userAgent = options.browser?.userAgent ?? options.userAgent ?? desktopUserAgent;
  const locale = options.browser?.locale ?? options.locale ?? "de-CH";

  const browser: PlaywrightContextProfile = {
    ...options.browser,
    userAgent,
    locale,
    viewport: options.browser?.viewport ?? { width: 1920, height: 1080 },
    stealth: options.browser?.stealth ?? true
  };

  return await withPlaywrightContext(browser, async (context) => {
    const page = await context.newPage();
    try {
      options.log?.(`Loading ${options.listingUrl}`);
      await page.goto(options.listingUrl, { waitUntil: "domcontentloaded", timeout: 90_000 });
      await handleCookieConsent(page);

    try {
      await page.waitForSelector('article:has(a[href*="/p/"])', { timeout: 60_000, state: "attached" });
    } catch {
      // allow empty listing
    }
    await page.waitForTimeout(1200);

    const raw = await page.$$eval('article:has(a[href*="/p/"])', (els: any[]) =>
      els.map((el) => {
        const a = el.querySelector?.('a[href*="/p/"]');
        const img = el.querySelector?.("img");
        const priceEl = el.querySelector?.('.vp-price [data-part="amount"], [data-part="amount"]');
        const url = a?.href ?? null;
        const nameFromImg = img?.alt ? String(img.alt).trim() : null;
        const nameFromLink = a?.textContent ? String(a.textContent).trim() : null;
        const srcset = img?.getAttribute?.("srcset") ?? img?.getAttribute?.("data-srcset") ?? null;
        const imgUrl = img?.currentSrc ?? img?.src ?? img?.getAttribute?.("data-src") ?? null;
        const priceText = priceEl?.textContent ? String(priceEl.textContent).trim() : null;
        return { url, name: nameFromImg || nameFromLink, priceText, imgUrl, srcset };
      })
    );

    const products: DiscoveredProduct[] = [];
    for (const item of raw) {
      const urlRaw = asNonEmptyString(item.url);
      const nameRaw = asNonEmptyString(item.name);
      if (!urlRaw || !nameRaw) continue;

      const url = urlRaw.split("?", 1)[0]!.trim();
      if (!url) continue;

      const itemId = extractItemId(urlRaw) ?? url;
      const { price, currency: parsedCurrency } = parsePriceText(
        typeof item.priceText === "string" ? item.priceText : undefined
      );
      const currency = options.currency ?? parsedCurrency ?? (price !== undefined ? "CHF" : undefined);

      const imageCandidate = asNonEmptyString(
        pickBestSrcsetUrl(typeof item.srcset === "string" ? item.srcset : undefined) ??
          (typeof item.imgUrl === "string" ? item.imgUrl : undefined)
      );
      const imageUrl = normalizeUrl(baseUrl, imageCandidate);

      products.push({
        sourceSlug: options.sourceSlug,
        itemId,
        name: normalizeSpaces(nameRaw),
        url,
        price,
        currency,
        imageUrl
      });
    }

      return { products };
    } finally {
      await page.close().catch(() => undefined);
    }
  });
}
